fix: propagate permissions issues portforward

SWITCHin2 opened the pull request:
fixes: https://github.com/kr8s-org/kr8s/issues/540

SATYAsasini committed:
feat: raise exception for unathenticated port-forward

SWITCHin2 committed:
Merge branch 'kr8s-org:main' into main

jacobtomlinson commented on kr8s/_portforward.py:
```diff
@@ -238,7 +239,18 @@ async def _connect_websocket(self):
+                        error_message = e.response.text
+                    raise ServerError(error_message, response=e.response) from e
                 self.pod = None
+                connection_attempts += 1
```
Nice catch, I think not incrementing this counter is the root of the problem

jacobtomlinson reviewed (commented):
This looks good thanks! 

Could I ask you to add a test that verifies this behaviour. I would start by finding a test that creates a temporary copy of the kubeconfig copy that. Then change token in the kubeconfig to something invalid and then trying to open a port forward

SWITCHin2 commented on kr8s/_portforward.py:
```diff
@@ -238,7 +239,18 @@ async def _connect_websocket(self):
+                    with suppress(httpx.StreamClosed, httpx.ResponseNotRead):
+                        await e.response.aread()
+                        error_message = e.response.text
+                    raise ServerError(error_message, response=e.response) from e
```
Ummm yes thats makes sense, sure will do that

SWITCHin2 commented:
> This looks good thanks!
> 
> Could I ask you to add a test that verifies this behaviour. I would start by finding a test that creates a temporary copy of the kubeconfig copy that. Then change token in the kubeconfig to something invalid and then trying to open a port forward

Sure, actually I had an testing script which does something like this will add the same in to /test dir

SATYAsasini committed:
fix: incorporate review comment and add test script

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

SATYAsasini committed:
merge conflict resolve

SATYAsasini committed:
fix: add end comment in test script

SWITCHin2 commented:
hey @jacobtomlinson 

- [x] have handled premature error throwing case as you mentioned
- [x] modified my test script as per you said , pushed them and also fixed some pre checks

jacobtomlinson reviewed (changes requested):
Thanks for pushing the test. I left a few comments but overall there is a lot of unnecessary code in there.

I think you just need to:
- Create the temporary credentials with bad token
- Create the `api_invalid` 
- Create the invalid Pod object (but not actually create the resource)
- Try to call the port forward and assert that it raises the `ServerError`

SATYAsasini committed:
fix: remove unneccesary codes in test file

SWITCHin2 commented on kr8s/tests/test_portforward_auth.py:
```diff
@@ -0,0 +1,160 @@
+        def kubeconfig_path(self):
+            return Path(os.environ.get("KUBECONFIG", "~/.kube/config")).expanduser()
+
+    return MockCluster()
```
yes I was using this for my local dev test, 
sure have incorporated that

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

SATYAsasini committed:
fix: linting isue and unused import

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jacobtomlinson commented on kr8s/tests/test_portforward_auth.py:
```diff
@@ -0,0 +1,64 @@
+    try:
+        api_invalid = await kr8s.asyncio.api(kubeconfig=kubeconfig_data)
+    finally:
+        Api._check_version = original_check_version
```
Using the mock patch context manager would hugely simplify this.

jacobtomlinson reviewed (commented):
A couple more comments on the tests, but otherwise this looks good to go.

Could I also ask you to put these tests in the existing `test_objects.py` file next to the other port forward tests instead of creating a new one. AI code tools love to create new test files unnecessarily.

SATYAsasini committed:
fix: add portforward to test_object.py and resolve some review comments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

SATYAsasini committed:
misc: remove test portward auth test file

SATYAsasini committed:
Merge branch 'main' of https://github.com/SWITCHin2/kr8s

SWITCHin2 committed:
Merge branch 'kr8s-org:main' into main

SATYAsasini committed:
fix: off-by-one error in portforward retry logic

SWITCHin2 committed:
Merge branch 'kr8s-org:main' into main

SWITCHin2 commented:
@jacobtomlinson I tried running the CI tests on my forked repo from the latest main branch, but i think there are multiple failed cases related to `test_pod_exec_error` and `test_pod_exec_to_file`. Could you please verify if it is for all or something is issue from my side

SATYAsasini committed:
fix: raise ServerError only after retries exhausted, fix py3.9 test mock

Move the permission-error (ServerError) branch inside the retries-exhausted
block so token rotation isn't short-circuited, per review.

Fix test_portforward_invalid_token on Python 3.9: patch the Api class object
directly instead of the string target 'kr8s._api.Api', which 3.9's stdlib mock
cannot resolve because the kr8s._api name is shadowed by the api function.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jacobtomlinson committed:
Merge branch 'main' of github.com:kr8s-org/kr8s into SWITCHin2/main

SWITCHin2 commented:
@jacobtomlinson thank you for triggerring checks, i found some issues from my side and have fixed them, but the `test_pod_exec_error` and `test_pod_exec_to_file` things seems still there, do let me know the update then i will tewt my changes and let you know when it will be ready to merge

jacobtomlinson commented:
I've unblocked CI in #753 and merged those into this PR so CI should be happier now.

jacobtomlinson commented:
Although GitHub seems to be having a partial outage, so things aren't perfect right now 😅.

SWITCHin2 commented:
Thanks, wrapping up things from myside

jacobtomlinson commented:
CI is running now

SWITCHin2 commented:
@jacobtomlinson all tests are passed but the `pre-commit.ci ` has failed while waiting for mergeable check, so can you re trigger it or should i do a empty commit to trigger all

jacobtomlinson commented:
Merged! Thanks for your patience here, I've been busy with other things lately and haven't given kr8s enough love.

SWITCHin2 commented:
Thanks @jacobtomlinson, That's completely understandable. The community has got kr8s covered.
